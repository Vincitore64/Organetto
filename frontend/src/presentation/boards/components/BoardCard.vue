<template>
  <a-card hoverable class="board-card" :body-style="{ padding: '16px' }" @click="$emit('open', board.id)">
    <template #cover>
      <div class="card-cover">
        <img :src="board.thumbnailUrl" :alt="board.title" class="board-thumbnail" />
        <div class="overlay">
          <div class="overlay-content">
            <EyeOutlined class="view-icon" />
            <span class="view-text">Открыть доску</span>
          </div>
        </div>
      </div>
    </template>

    <div class="card-content">
      <h3 class="board-title">{{ board.title }}</h3>
      <div class="board-meta">
        <div class="meta-item">
          <ClockCircleOutlined class="meta-icon" />
          <span class="meta-text">Недавно</span>
        </div>
        <div class="meta-item">
          <UserOutlined class="meta-icon" />
          <span class="meta-text">Личная</span>
        </div>
      </div>
    </div>
  </a-card>
</template>

<script setup lang="ts">
import { EyeOutlined, ClockCircleOutlined, UserOutlined } from '@ant-design/icons-vue'

interface Board {
  id: number
  title: string
  thumbnailUrl: string
}

defineProps<{ board: Board }>()
defineEmits<{
  open: [id: number]
}>()
</script>

<style scoped lang="scss">
.board-card {
  display: grid;
  grid-template-rows: 1fr auto;
  border-radius: 12px;
  border: 1px solid rgba(0, 0, 0, 0.06);
  background: var(--color-surface, #ffffff);
  transition: all 0.3s cubic-bezier(0.4, 0, 0.2, 1);
  cursor: pointer;
  overflow: hidden;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.06);

  &:hover {
    transform: translateY(-4px);
    box-shadow: 0 8px 32px rgba(0, 0, 0, 0.12);
    border-color: var(--color-primary-200, #bae7ff);

    .overlay {
      opacity: 1;
    }

    .board-thumbnail {
      transform: scale(1.05);
    }
  }

  :deep(.ant-card-body) {
    padding: 16px;
  }
}

.card-cover {
  position: relative;
  height: 100%;
  overflow: hidden;
  background: linear-gradient(135deg, #f0f2f5, #e6f7ff);

  .board-thumbnail {
    width: 100%;
    height: 100%;
    object-fit: cover;
    transition: transform 0.3s cubic-bezier(0.4, 0, 0.2, 1);
  }

  .overlay {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background: rgba(0, 0, 0, 0.6);
    display: flex;
    align-items: center;
    justify-content: center;
    opacity: 0;
    transition: opacity 0.3s cubic-bezier(0.4, 0, 0.2, 1);

    .overlay-content {
      display: flex;
      flex-direction: column;
      align-items: center;
      gap: 8px;
      color: white;

      .view-icon {
        font-size: 24px;
      }

      .view-text {
        font-size: 16px;
        font-weight: 500;
      }
    }
  }
}

.card-content {
  .board-title {
    font-size: 18px;
    font-weight: 600;
    color: var(--color-text, #262626);
    margin: 0 0 12px 0;
    line-height: 1.4;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .board-meta {
    display: flex;
    gap: 16px;

    .meta-item {
      display: flex;
      align-items: center;
      gap: 4px;

      .meta-icon {
        font-size: 14px;
        color: var(--color-text-weak, #8c8c8c);
      }

      .meta-text {
        font-size: 14px;
        color: var(--color-text-weak, #8c8c8c);
        font-weight: 400;
      }
    }
  }
}
</style>