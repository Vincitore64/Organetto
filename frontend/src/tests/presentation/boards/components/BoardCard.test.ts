import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import BoardCard from '@/presentation/boards/components/BoardCard.vue'
import type { BoardPageView } from '@/presentation/boards/models'
// import { defaultThumbnailHref } from '../models'

// Mock Ant Design Vue icons
const mockIcons = {
  EyeOutlined: { template: '<span class="eye-icon">👁</span>' },
  ClockCircleOutlined: { template: '<span class="clock-icon">🕐</span>' },
  UserOutlined: { template: '<span class="user-icon">👤</span>' },
}

const createMockBoard = (overrides: Partial<BoardPageView> = {}): BoardPageView => ({
  id: 1,
  title: 'Test Board',
  thumbnailUrl: '/test-image.png',
  ...overrides,
})

describe('BoardCard.vue', () => {
  // 1. Rendering tests
  describe('Rendering', () => {
    it('renders the board title', () => {
      const board = createMockBoard({ title: 'My Board' })
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })
      expect(wrapper.text()).toContain('My Board')
    })

    it('renders the board thumbnail with correct src', () => {
      const board = createMockBoard({ thumbnailUrl: '/custom-image.png' })
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })
      const img = wrapper.find('.board-thumbnail')
      expect(img.attributes('src')).toBe('/custom-image.png')
    })

    it('renders the board thumbnail with correct alt text', () => {
      const board = createMockBoard({ title: 'My Board Title' })
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })
      const img = wrapper.find('.board-thumbnail')
      expect(img.attributes('alt')).toBe('My Board Title')
    })

    it('renders the overlay with view text', () => {
      const board = createMockBoard()
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })
      expect(wrapper.find('.overlay').exists()).toBe(true)
      expect(wrapper.find('.view-text').text()).toBe('Открыть доску')
    })

    it('renders meta information', () => {
      const board = createMockBoard()
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })
      expect(wrapper.find('.board-meta').exists()).toBe(true)
      expect(wrapper.text()).toContain('Недавно')
      expect(wrapper.text()).toContain('Личная')
    })
  })

  // 2. Interaction tests
  describe('Interactions', () => {
    it('emits "open" event with board id when card is clicked', async () => {
      const board = createMockBoard({ id: 42 })
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      await wrapper.find('.board-card').trigger('click')

      expect(wrapper.emitted()).toHaveProperty('open')
      expect(wrapper.emitted('open')![0]).toEqual([42])
    })

    it('shows overlay on hover (CSS behavior)', () => {
      const board = createMockBoard()
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      const overlay = wrapper.find('.overlay')
      expect(overlay.exists()).toBe(true)
      // Note: CSS hover effects can't be directly tested in unit tests
      // but we can verify the overlay element exists
    })
  })

  // 3. Props validation
  describe('Props', () => {
    it('handles board prop correctly', () => {
      const board = createMockBoard({
        id: 123,
        title: 'Custom Title',
        thumbnailUrl: '/custom.jpg',
      })

      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      expect(wrapper.find('.board-title').text()).toBe('Custom Title')
      expect(wrapper.find('.board-thumbnail').attributes('src')).toBe('/custom.jpg')
    })

    it('truncates long titles with CSS ellipsis', () => {
      const longTitle = 'This is a very long board title that should be truncated with ellipsis'
      const board = createMockBoard({ title: longTitle })

      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      const titleElement = wrapper.find('.board-title')
      expect(titleElement.text()).toBe(longTitle)
      // CSS truncation is handled by white-space: nowrap and text-overflow: ellipsis
      expect(titleElement.classes()).not.toContain('truncated')
    })
  })

  // 4. Edge cases
  describe('Edge Cases', () => {
    it('handles empty thumbnail URL gracefully', () => {
      const board = createMockBoard({ thumbnailUrl: '' })
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      const img = wrapper.find('.board-thumbnail')
      expect(img.attributes('src')).toBe('')
    })

    it('handles special characters in title', () => {
      const board = createMockBoard({ title: 'Board with <script>alert("xss")</script> & symbols' })
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      expect(wrapper.find('.board-title').text()).toContain('Board with')
      expect(wrapper.find('.board-title').text()).toContain('& symbols')
    })

    it('handles zero or negative board id', async () => {
      const board = createMockBoard({ id: -1 })
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      await wrapper.find('.board-card').trigger('click')
      expect(wrapper.emitted('open')![0]).toEqual([-1])
    })
  })

  // 5. Component structure
  describe('Component Structure', () => {
    it('has correct CSS classes', () => {
      const board = createMockBoard()
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      expect(wrapper.find('.board-card').exists()).toBe(true)
      expect(wrapper.find('.card-cover').exists()).toBe(true)
      expect(wrapper.find('.card-content').exists()).toBe(true)
      expect(wrapper.find('.board-title').exists()).toBe(true)
      expect(wrapper.find('.board-meta').exists()).toBe(true)
    })

    it('uses Ant Design Card component', () => {
      const board = createMockBoard()
      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: {
            ...mockIcons,
            'a-card': {
              template: '<div class="ant-card"><slot name="cover"></slot><slot></slot></div>',
              props: ['hoverable', 'bodyStyle'],
            },
          },
        },
      })

      expect(wrapper.find('.ant-card').exists()).toBe(true)
    })
  })

  // 6. Snapshot test
  describe('Snapshot', () => {
    it('matches snapshot for a normal board', () => {
      const board = createMockBoard({
        id: 1,
        title: 'Sample Board',
        thumbnailUrl: '/sample.png',
      })

      const wrapper = mount(BoardCard, {
        props: { board },
        global: {
          components: mockIcons,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
