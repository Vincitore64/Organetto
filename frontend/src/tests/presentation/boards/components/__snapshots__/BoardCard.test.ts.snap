// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BoardCard.vue > Snapshot > matches snapshot for a normal board 1`] = `
"<a-card data-v-14f40d1b="" hoverable="" class="board-card" body-style="[object Object]">
  <div data-v-14f40d1b="" class="card-content">
    <h3 data-v-14f40d1b="" class="board-title">Sample Board</h3>
    <div data-v-14f40d1b="" class="board-meta">
      <div data-v-14f40d1b="" class="meta-item"><span data-v-14f40d1b="" role="img" aria-label="clock-circle" class="anticon anticon-clock-circle meta-icon"><svg focusable="false" data-icon="clock-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"></path><path d="M686.7 638.6L544.1 535.5V288c0-4.4-3.6-8-8-8H488c-4.4 0-8 3.6-8 8v275.4c0 2.6 1.2 5 3.3 6.5l165.4 120.6c3.6 2.6 8.6 1.8 11.2-1.7l28.6-39c2.6-3.7 1.8-8.7-1.8-11.2z"></path></svg><!----></span><span data-v-14f40d1b="" class="meta-text">Недавно</span></div>
      <div data-v-14f40d1b="" class="meta-item"><span data-v-14f40d1b="" role="img" aria-label="user" class="anticon anticon-user meta-icon"><svg focusable="false" data-icon="user" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"></path></svg><!----></span><span data-v-14f40d1b="" class="meta-text">Личная</span></div>
    </div>
  </div>
</a-card>"
`;
